### ISSUE TITLE
[MNG-2140] Mojo in forked lifecycle invoked in reactor build cannot find reactor dependency

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#4309]
**[Grzegorz Slowikowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gslowikowski)** opened **[MNG-2140](https://issues.apache.org/jira/browse/MNG-2140?redirect=false)** and commented

In multiproject I have two modules. One of them (lets name it "main") depends on the second one ("util").
If I bind forking lifecycle mojo in the "main" artifact pom, that mojo does not find "util" artifact and tries to find it in local and remote repositories.
If the same mojo does not for lifecycle, reactor build finishes with success.

I attach an example:
- plugin with two mojos writing something to the log, one of them forks lifecycle, second - does not
- test multiproject using test mojo

There are some plugins forking lifecycles, so I think this can be serious problem.
I would love to see its priority to "Blocker" for Maven 2.0.3

---

**Affects:** 2.0.2

**Attachments:**
- [test.zip](https://issues.apache.org/jira/secure/attachment/12711749/test.zip) (_4.50 kB_)

**Issue Links:**
- [MNG-1682](https://issues.apache.org/jira/browse/MNG-1682) Plugins do not honor the correct extension when run as a part of a multiproject build

1 votes, 2 watchers

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

not blocking 2.0.3. It's not a regression.

The workaround is to "mvn install" instead of "mvn package". I'd suggest that the forked lifecycle is not cloning the list of already resolved projects and their locations.

**Comment by jira-importer** [comment]
**[Walco van Loon](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=walco)** commented

Like the original reporter, I think it's important to fix this issue, because it will prevent plugins like Cobertura and Clover to run in a multi project build if the modules have interdependencies. For these plugins, the suggested workaround is not really convenient, as they are supposed to execute in a phase prior to the install phase.

**Comment by jira-importer** [comment]
**[Guillaume Duchesneau](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gduchesneau)** commented

I am using the Cobertura plugin and as Walco said in its comment, it is not very convenient to install all modules prior to launching the tests for code coverage...

To understand what was going on behind the scene in Maven, I wrote two simple test mojos : one that forks the lifecycle and one that does not.  I found out that the field projectReferences in the class org.apache.maven.project.MavenProject is not set when forking the lifecycle causing the artifact resolution to fail for project in the reactor not yet installed.  I (maybe naively) think that the problem is simply due to the fact that the copy constructor in MavenProject does not copy that field.  I used Maven 2.0.4 sources to step in the code.

I hope this can help...
